traffic_ops/traffic_ops_golang/deliveryservice/deliveryservicesv13.go

```diff
@@ -0,0 +1,1109 @@
+		return tc.DBError, tc.SystemError
+	} else if !ok {
+		log.Errorln("TODeliveryServiceV13.Delete ID '" + string(*tods.DS.ID) + "' had no delivery service!")
+		return tc.DBError, tc.DataMissingError
+	} else if ds.XMLID == nil {
+		// should never happen
+		log.Errorf("TODeliveryServiceV13.Delete ID '%v' loaded nil XML ID!", *tods.DS.ID)
+		return tc.DBError, tc.SystemError
+	}
+
+	result, err := tx.Exec(`DELETE FROM deliveryservice WHERE id=$1`, ds.ID)
+	if err != nil {
+		log.Errorln("TODeliveryServiceV13.Delete deleting delivery service: " + err.Error())
+		return tc.DBError, tc.SystemError
+	}
+	rowsAffected, err := result.RowsAffected()
+	if err != nil {
+		return tc.DBError, tc.SystemError
+	}
+	if rowsAffected != 1 {
+		if rowsAffected < 1 {
+			return errors.New("no delivery service with that id found"), tc.DataMissingError
+		}
+		return fmt.Errorf("this create affected too many rows: %d", rowsAffected), tc.SystemError
+	}
+
+	if _, err := tx.Exec(`DELETE FROM deliveryservice_regex WHERE deliveryservice=$1`, ds.ID); err != nil {
+		log.Errorln("TODeliveryServiceV13.Delete deleting delivery service regexes: " + err.Error())
+		return tc.DBError, tc.SystemError
+	}
+
+	paramConfigFilePrefixes := []string{"hdr_rw_", "hdr_rw_mid_", "regex_remap_", "cacheurl_"}
+	configFiles := []string{}
+	for _, prefix := range paramConfigFilePrefixes {
+		configFiles = append(configFiles, prefix+*ds.XMLID+".config")
+	}
+
+	if _, err := tx.Exec(`DELETE FROM parameter WHERE name = 'location' AND config_file = ANY($1)`, pq.Array(configFiles)); err != nil {
+		log.Errorln("TODeliveryServiceV13.Delete deleting delivery service parameters: " + err.Error())
+		return tc.DBError, tc.SystemError
+	}
+
+	commitTx = true
+	return nil, tc.NoError
+}
+
+// IsTenantAuthorized implements the Tenantable interface to ensure the user is authorized on the deliveryservice tenant
+func (tods *TODeliveryServiceV13) IsTenantAuthorized(user auth.CurrentUser, db *sqlx.DB) (bool, error) {
+	tods.V12().LoadTenantID(db) // ignore errors and keep user-set tenant ID (which happens on Create)
+
+	// TODO make Tenant framework able to handle loading vs accepting user input, for Create vs Update vs Delete
+
+	// if ok, err := tods.V12().LoadTenantID(db); err != nil {
+	// 	return false, errors.New("loading tenant ID from database: " + err.Error())
+	// } else if !ok {
+	// 	return false, errors.New("delivery service not found in database")
+	// }
+	if tods.DS.TenantID == nil {
+		log.Debugf("tenantID is nil")
+		return false, errors.New("tenantID is nil")
+	}
+	return tenant.IsResourceAuthorizedToUser(*tods.DS.TenantID, user, db)
+}
+
+func filterAuthorized(dses []tc.DeliveryServiceNullableV13, user auth.CurrentUser, db *sqlx.DB) ([]tc.DeliveryServiceNullableV13, error) {
+	newDSes := []tc.DeliveryServiceNullableV13{}
+	for _, ds := range dses {
+		if ds.TenantID == nil {
+			if ds.XMLID == nil {
```

Comment:
Can this ever actually be true since the xml_id column is not-null?

Reply:
Yes, because the function may be given objects which were user input, not loaded from the database.

Reply:
ah, thanks for the clarification. So it could be given objects which were user input, but currently it's only used for objects loaded from the db. Gotcha